### ISSUE TITLE
[Feature] OpenSearch and Apache Spark Integration

### ISSUE THREAD

**Issue by penghuo** [opensearch-project/opensearch-spark#3]
More reading at https://github.com/opensearch-project/opensearch-spark/issues/4

### Goals

* Verify the solution for end to end production use case
* Demo the solution

### Tasks

- [x] Use Case Analysis
    - [x] MV definition: build desired dashboard in Superset/Grafana and figure out MV required to accelerate the queries
    - [x] Spark transform logic: build MV by Spark DataFrame and verify if any functional gap in Spark streaming
- [x] Setup Spark docker cluster
    - [x] Thrfit JDBC Server and LDAP configuration *[Deprecated]*
    - [x] opensearch-project/sql#1352
- [x] Backend Development
    - [x] Add OpenSearch Spark Connector
        - [x] https://github.com/opensearch-project/sql/issues/1302
        - [x] https://github.com/opensearch-project/sql/issues/1306
        - [x] https://github.com/opensearch-project/sql/pull/1311
        - [x] https://github.com/opensearch-project/opensearch-hadoop/pull/48
    - [x] Maximus Implementation
        - [x] opensearch-project/sql#1363
            - [x] opensearch-project/sql#1394
            - [x] opensearch-project/sql#1362
        - [x] Store Maximus Table metadata inside OpenSearch.
            - [x] store metadata/index/mv in opensearch.
            - [x] sync between table and index/mv *[Deprioritized]*
        - [x] https://github.com/opensearch-project/sql/issues/1379
            - [x] Extend Spark grammar with `CREATE INDEX`
            - [x] Store index data in OpenSearch index
            - [x] Auto refresh by Spark streaming execution (integrate with table metadata refresh job)
        - [x] https://github.com/opensearch-project/sql/issues/1407
            - [x] Extend Spark grammar with `CREATE MATERIALIZED VIEW`
            - [x] Refresh MV data by Spark streaming job
            - [x] Add tumbling window function
        - [x] https://github.com/opensearch-project/sql/issues/1435
            - [x] Rewrite query to use skipping index (reuse Hyperspace)
            - [x] Rewrite streaming query (create MV) with skipping index
            - [x] Rewrite direct query with materialized view *[Deprioritized]*
- [x] Demo
    - [x] Create dashboard in Superset/Grafana for demo use
    - [x] Integrate and demo with OpenSearch Observability once the UI is ready
- [x] Release 
    - [x] opensearch-project/sql#1480

### Deliverables

* Docker: https://github.com/penghuo/os-sql/tree/feature/spark-integration/docker#setup-dev-env
* Demo: https://github.com/opensearch-project/sql/discussions/1465

---
### Goals

* OpenSearch Release: 2.9.0
  * Support create spark-submit datasource in OpenSearch.
  * Support build visualization in OpenSearch Dashboard using spark datasource.
* OpenSearch Spark Extension Release. 
  * Support create skipping index for Hive Table. 

```
// 1. define the context
USING FLINT
PROPERTIES (
  OPENSEARCH = "https://my-opensearch.com"
)

// 2. create skipping index
CREATE SKIPPING INDEX index_name ON TABLE [db_name].table_name
FOR COLUMNS (col1 INDEX_TYPE, ...)
[REFRESH_ON_CREATE]            <- auto refresh the table after table create
[AUTO_REFRESH TRUE]            <- auto refresh the index when new data appended.
```

### Tasks

- [x] Backend Development
    - [x] Codebase
        - [x] https://github.com/opensearch-project/sql/pull/1573
    - [x] Spark Connector in OpenSearch SQL
        - [x] https://github.com/opensearch-project/sql/pull/1780
        - [x] https://github.com/opensearch-project/sql/pull/1723
    - [x] https://github.com/opensearch-project/sql/issues/1635
        - [x] https://github.com/opensearch-project/sql/pull/1653
        - [x] https://github.com/opensearch-project/sql/pull/1633
        - [x] https://github.com/opensearch-project/sql/pull/1634
        - [x] https://github.com/opensearch-project/sql/pull/1663
    - [x] Flint Spark Integration
        - [x] Flint Connector
            - [x] Spark SPI
                - [x] opensearch-project/sql#1631
                    - [x] opensearch-project/sql#1632
                    - [x] Partition Support
                    - [x] opensearch-project/sql#1641
                    - [x] https://github.com/opensearch-project/sql/pull/1699
                - [x] https://github.com/opensearch-project/sql/pull/1656
            - [x] OpenSearch Sink
                - [x] Improve schema infer in OpenSearch-Hadoop
        - [x] https://github.com/opensearch-project/opensearch-spark/issues/2
            - [x] Skipping Index SQL Support
                - [x] Flint context configuration statement
                - [x] https://github.com/opensearch-project/sql/pull/1800
                - [x] https://github.com/opensearch-project/sql/pull/1785
                - [x] https://github.com/opensearch-project/sql/pull/1730
            - [x] Skipping Index API Interface
                - [x] Create/delete/show Index
                    - [x] https://github.com/opensearch-project/sql/pull/1636
                - [x] Refresh Index
                    - [x] Add `REFRESH INDEX` statement
                    - [x] https://github.com/opensearch-project/sql/pull/1658
            - [x] Skipping Index Provider
                - [x] Partition index
                - [x]  https://github.com/opensearch-project/sql/pull/1708
                - [x]  https://github.com/opensearch-project/sql/pull/1726
            - [x] Spark query optimization
                - [x] https://github.com/opensearch-project/sql/pull/1690
    - [x] Fault Tolerance
        - [x] Streaming job can resume with checkpoint
        - [x] https://github.com/opensearch-project/sql/pull/1803
        - [x] https://github.com/opensearch-project/sql/pull/1819
    - [x] Bug Fixes, Enhancements, Refactoring
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/8
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/7
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/10
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/9
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/20

### Deliverables

* Demo (without OS Spark connector): https://github.com/opensearch-project/sql/discussions/1877

---

### Goals

* OpenSearch release: preview in 2.11
* User story: https://github.com/opensearch-project/sql/issues/1376
* User guide: guidance for performance, cost estimation and OpenSearch/Spark cluster sizing

### Non-Goals

* Covering index and MV are not accessible in Spark and thus no query rewrite support

### Tasks
- [x] Backend Development
    - [x] https://github.com/opensearch-project/opensearch-spark/issues/23
        - [x] API support
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/22
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/143 *[TBD]*
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/85
            - [x] Support included columns *[Low priority]*
        - [x] SQL support
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/32
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/36
    - [x] https://github.com/opensearch-project/opensearch-spark/issues/25
        - [x] API support
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/71
            - [x] Handle large backlog when cold start *[Postponed]*
        - [x] SQL support
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/73
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/81
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/97
        - [x] Windowing Function
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/37
    - [x] https://github.com/opensearch-project/opensearch-spark/issues/57
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/58
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/82
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/119
    - [x] https://github.com/opensearch-project/opensearch-spark/issues/95
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/110
    - [x] Bug Fixes, Enhancements, Refactoring
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/17
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/21
        - [x] https://github.com/opensearch-project/opensearch-spark/issues/24
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/70
            - [x] Build API refactor is done in MV API PR
        - [x] https://github.com/opensearch-project/opensearch-spark/issues/26
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/39
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/44
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/76
            - [x] Separate auto start option from auto refresh *[Low priority]*
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/42
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/47
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/48
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/49
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/60
        - [x] https://github.com/opensearch-project/opensearch-spark/issues/65
            - [x] https://github.com/opensearch-project/opensearch-spark/pull/66 
        - [x] https://github.com/opensearch-project/opensearch-spark/issues/53
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/92
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/98
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/125
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/127
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/142
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/158
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/169
        - [x] https://github.com/opensearch-project/opensearch-spark/pull/170
- [x] Frontend Development
    - [x] Table, index and MV management UI which allow user manage the table, index and MV.
    - [x] ~SQL explain UI which allow user to know (1) the benefit of create index or MV and (2) the query execution plan and the cost of each operation.~
- [x] Performance Test
    - [x] opensearch-project/sql#1400

**Comment by dai-chen** [comment]
## Phase-3: Flint Index Enhancement and Optimization

### Goals

* OpenSearch release: TBD
* User story: TBD
* User guide: guidance for performance, cost estimation and OpenSearch/Spark cluster sizing

### Non-Goals

* N/A

### Tasks
- [ ] https://github.com/opensearch-project/opensearch-spark/issues/118
    - [ ] Automatic skipping algorithm selection
    - [ ] Lazy skipping index building
    - [ ] Persist skipping index data to S3
- [ ] https://github.com/opensearch-project/opensearch-spark/issues/29
- [ ] https://github.com/opensearch-project/opensearch-spark/issues/104
- [ ] Testing
    - [ ] Documentation
        - [ ] https://github.com/opensearch-project/opensearch-spark/issues/28
    - [ ] Performance
        - [ ] https://github.com/opensearch-project/opensearch-spark/issues/38
        - [ ] https://github.com/opensearch-project/sql/issues/1399
        - [ ] https://github.com/opensearch-project/sql/issues/1401
        - [ ] https://github.com/opensearch-project/sql/issues/1402

**Comment by dai-chen** [comment]
Follow-up tracking issue: https://github.com/opensearch-project/opensearch-spark/issues/365
